### PR TITLE
[SPARK-17219][ML] Add NaN value handling in Bucketizer

### PR DESCRIPTION
**Pull request by VinceShieh**
## What changes were proposed in this pull request?

This PR fixes an issue when Bucketizer is called to handle a dataset containing NaN value.
Sometimes, null value might also be useful to users, so in these cases, Bucketizer should
reserve one extra bucket for NaN values, instead of throwing an illegal exception.
Before:

```
Bucketizer.transform on NaN value threw an illegal exception.
```

After:

```
NaN values will be grouped in an extra bucket.
```
## How was this patch tested?

New test cases added in `BucketizerSuite`.
Signed-off-by: VinceShieh vincent.xie@intel.com

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala [hunk 1]
```diff
@@ -129,17 +129,21 @@ object Bucketizer extends DefaultParamsReadable[Bucketizer] {
     if (feature == splits.last) {
       splits.length - 2
     } else {
-      val idx = ju.Arrays.binarySearch(splits, feature)
-      if (idx >= 0) {
-        idx
+      if (feature.isNaN) {
+        splits.length - 1
```
Comment (VinceShieh) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: index NaN value as ${splits.length -1}
''' all the other changes within this files, without comments, are just for code refactoring

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala [hunk 2]
```diff
@@ -63,7 +63,7 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
 
   /** @group setParam */
   @Since("1.4.0")
-  def setSplits(value: Array[Double]): this.type = set(splits, value)
+  def setSplits(value: Array[Double]): this.type = set(splits, value.filter(!_.isNaN))
```
Comment (VinceShieh) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: we only accept non-Nan splits

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: How about just rejecting splits with NaN? I don't see a valid reason a caller could specify splits bounded by NaN since it has no ordering.

File: mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala [hunk 3]
```diff
@@ -116,8 +116,7 @@ final class QuantileDiscretizer @Since("1.6.0") (@Since("1.6.0") override val ui
 
     val distinctSplits = splits.distinct
     if (splits.length != distinctSplits.length) {
-      log.warn(s"Some quantiles were identical. Bucketing to ${distinctSplits.length - 1}" +
-        s" buckets as a result.")
+      log.warn(s"Some quantiles were identical. Will return fewer buckets than requested!")
```
Comment (VinceShieh) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: considering cases when, there are duplicated non-nan values, along with multiple NaNs in the splits, to get the exact number of buckets, will require extra computation and increase complexity which I dont think it worths for a warning msg

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala (hunk 2 above)
Comment (VinceShieh) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: rejecting? do you mean we throw an error for splits with NaN ?
User might not specify NaN bounded splits, but it could be set by other function calls, such as approxQuantile, from a dataset with quite a number of NaN

File: mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala (hunk 3 above)
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: I think this message was OK?

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala [hunk 4]
```diff
@@ -129,17 +129,21 @@ object Bucketizer extends DefaultParamsReadable[Bucketizer] {
     if (feature == splits.last) {
       splits.length - 2
     } else {
-      val idx = ju.Arrays.binarySearch(splits, feature)
-      if (idx >= 0) {
-        idx
+      if (feature.isNaN) {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: I think this might just be me, but, it felt like the first two checks were at the same level:

```
if (...) {
} else if (...) {
} else {
  ...
}
```

Or alternatively, I could see handling the `feature == splits.last` case in the `idx >= 0` case. Really that's the case that handle exact match of a value and right now it's split across two clauses. Then it makes sense to start the method by handling NaN alone.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala (hunk 2 above)
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: Yeah, that seems like an error. NaN has no ordering so it doesn't make sense to consider or return them as part of quantiles. I think these other classes have to handle this case too and they can be rejected here. It's not something this class can handle for them because it doesn't know what the quantiles would have been without NaN.

File: mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala (hunk 3 above)
Comment (VinceShieh) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: when the splits has NaN, we have to know the number of non-NaN value in the splits to get the exact number of returned buckets, considering splits :{ -Inf, -1, 1, Inf, NaN, NaN}, the old msg will actually get us a wrong result

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: Yes, another reason to reject NaN splits.

File: mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala [hunk 5]
```diff
@@ -114,10 +114,10 @@ final class QuantileDiscretizer @Since("1.6.0") (@Since("1.6.0") override val ui
     splits(0) = Double.NegativeInfinity
     splits(splits.length - 1) = Double.PositiveInfinity
 
-    val distinctSplits = splits.distinct
+    val distinctSplits = splits.filter(!_.isNaN).distinct
     if (splits.length != distinctSplits.length) {
-      log.warn(s"Some quantiles were identical. Bucketing to ${distinctSplits.length - 1}" +
-        s" buckets as a result.")
+      log.warn(s"Bucketizing to ${distinctSplits.length - 1} buckets if some quantiles" +
+        s" were identical; or ${distinctSplits.length} buckets if NaN value existed")
```
Comment (VinceShieh) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: if we are to give user the exact number of returned buckets, we need to go through the whole input dataset to check whether NaN value exists, the computation incurred just for a log warning msg is too high, so I choose to give user such msg instead

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala [hunk 6]
```diff
@@ -106,18 +106,19 @@ final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String
...
+    * No NaN split should be accepted. */
   private[feature] def checkSplits(splits: Array[Double]): Boolean = {
     if (splits.length < 3) {
       false
     } else {
       var i = 0
       val n = splits.length - 1
       while (i < n) {
-        if (splits(i) >= splits(i + 1)) return false
+        if (splits(i) >= splits(i + 1) || splits(i).isNaN) return false
```
Comment (VinceShieh) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: add a safe checker for NaN split here

File: mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala (hunk 5 above)
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: I think this is now a bit confusing, since it's reporting different things based on state that isn't logged for the user. If it's hard just say "bucketing to fewer buckets" as before at this stage.

Comment (VinceShieh) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: I think we just put it "the returned number of buckets might differ from what was requested depending on the data sample values. " , since the result number could be less than/equal to the requested number when same quantiles were spotted. And if no same quantiles exists in the splits, but dataset has NaN value, the actually number of buckets would then be  greater than requested.

File: mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala [hunk 7]
```diff
@@ -114,11 +114,9 @@ final class QuantileDiscretizer @Since("1.6.0") (@Since("1.6.0") override val ui
     splits(0) = Double.NegativeInfinity
     splits(splits.length - 1) = Double.PositiveInfinity
 
-    val distinctSplits = splits.distinct
-    if (splits.length != distinctSplits.length) {
-      log.warn(s"Some quantiles were identical. Bucketing to ${distinctSplits.length - 1}" +
-        s" buckets as a result.")
-    }
+    val distinctSplits = splits.filter(!_.isNaN).distinct
+    log.info(s"The exact number of returned buckets might differ from what was requested" +
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: OK, though now this is logged unconditionally. I think we'd only want to log (an info-level message) if the number of buckets didn't match the request, which is what you had previously?

Separately, I think the behavior of NaN has to be documented somewhere in this class too, to make people aware that it's _always_ possible to get an extra bucket of data if there are NaNs.

Otherwise looking good to me.

**Comment by VinceShieh** [comment]
updated tests and documents related to this change

File: docs/ml-features.md [hunk 8]
```diff
@@ -1102,7 +1102,8 @@ for more details on the API.
 ## QuantileDiscretizer
 
 `QuantileDiscretizer` takes a column with continuous features and outputs a column with binned
-categorical features. The number of bins is set by the `numBuckets` parameter.
+categorical features. The number of bins is set by the `numBuckets` parameter, but it is
```
Comment (srowen) [inline docs/ml-features.md]: OK, but this doesn't specify the behavior. It should be explicit that while data will go into buckets 0 through numBuckets-1, that NaN values will be counted in bucket numBuckets.

File: mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala [hunk 9]
```diff
@@ -114,10 +115,10 @@ final class QuantileDiscretizer @Since("1.6.0") (@Since("1.6.0") override val ui
     splits(0) = Double.NegativeInfinity
     splits(splits.length - 1) = Double.PositiveInfinity
 
-    val distinctSplits = splits.distinct
+    val distinctSplits = splits.filter(!_.isNaN).distinct
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: Ah, I think this is a little bit at odds with the intent. We need to filter NaN before the data goes to approxQuantile. They should have no input on the quantiles. Then there's no need to filter out NaN from splits.

The message below can then remain unchanged from what it was before. This message is not related to the behavior of NaN.

Comment (VinceShieh) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: I didnt take this approach, becoz it will increase the complexity, especially when the dataset is huge

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: It's a correctness issue though. You will compute the wrong splits. Why is it complex? just filter the input, it seems.

Comment (VinceShieh) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: hmm, if we filter NaN before the data, with size m, goes to approxQuantile, the increased complexity would be o(m), filter NaN in the splits would normally be more cheap, with number of splits far less than that of entire dataset.

File: docs/ml-features.md (hunk 8 above)
Comment (VinceShieh) [inline docs/ml-features.md]: consider situation, when a high proportion of duplicated data and/or NaN exist in a data sample, the exact number of buckets is hard to get, it could be less than/equal to/ more than 'numBuckets'. what we can be sure is that, the NaN value if existed will be grouped in the last bucket.

Comment (srowen) [inline docs/ml-features.md]: It's always possible to have less data than buckets. The problem here is that you might have enough non-NaN data, even, to properly determine distinct buckets, but fail to do so because of NaNs making some splits NaN. You'd end up with fewer splits than intended when you could have created all meaningful splits.

Comment (VinceShieh) [inline docs/ml-features.md]: for this part of documentation, how about put it:
"The number of bins is set by the `numBuckets` parameter, but the returned 'actualNumBuckets' might differ from what was request depending on the data sample value; while data will go into Buckets[0] to Buckets[actualNumBuckets - 1] and NaN value, if existed, will go into Buckets[actualNumBuckets]"
sounds good?

Comment (srowen) [inline docs/ml-features.md]: I would suggest:

The number of bins is set by the `numBuckets` parameter. It's possible that the number of buckets used will be less than this value, for example, if there are too few distinct values of the input to create enough distinct quantiles. Note also that NaN values are handled specially and placed into their own bucket. For example, if 4 buckets are used, then non-NaN data will be put into buckets 0-3, but NaNs will be counted in a special bucket 4.

Comment (VinceShieh) [inline docs/ml-features.md]: Nit. Thanks!

File: mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala (hunk 9 above)
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: Sure it could be faster, but it's also incorrect. We can't return a wrong answer here to save time. Do we disagree on the effect of including NaN when computing the quantiles?

Comment (VinceShieh) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: @srowen  then maybe we should, as we discussed earlier on JIRA, align with R, by having a NaN checker in approxQuantile, that is, having a NaN filter inside of approxQuantile, rather than ahead of calling approxQuantile.

We can also have a same flag for user to choose to either remove NaN values or throw an error when there is NaN in data,  although, this API change will introduce collateral impact on several existing function calls.

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: What would you do with NaN in approxQuantile except ignore it? it can't affect a quantile. The question of what the bucketizer should do with NaN is separate. I'd still favor one behavior there, but keeping that separate, let's figure out what approxQuantile should do.

**Comment by srowen** [comment]
Great, I have one last request @VinceShieh and that is to update the docs for QuantileDiscretizer in Scala and Python to reflect the additional comment about NaN that you put in the main docs. That would really complete it. I think the code and behavior looks solid now.

**Comment by VinceShieh** [comment]
@srowen Updated. Thanks.

File: mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala [hunk 10]
```diff
@@ -109,7 +114,7 @@ final class QuantileDiscretizer @Since("1.6.0") (@Since("1.6.0") override val ui
   @Since("2.0.0")
   override def fit(dataset: Dataset[_]): Bucketizer = {
     transformSchema(dataset.schema, logging = true)
-    val splits = dataset.stat.approxQuantile($(inputCol),
+    val splits = dataset.select($(inputCol)).na.drop().stat.approxQuantile($(inputCol),
```
Comment (VinceShieh) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: @srowen I have one question here. The reason why we dont move this NaN filter into approxiQuantile or multipleApproxQuantiles is that, those apis are shared with sparkSQL? Becoz, personally I think it would look better if we put this filter inside multipleApproxQuantiles, though it would introduce more changes and,  should make sure it doesnt impact other components other than mllib.

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: Yes, I agree that a similar argument applies for approxQuantile methods. I think the most reasonable semantics are to ignore NaN as well.

QuantileSummaries should probably reject insertion of NaN too.

I'd support making that change as well here, and expanding the scope accordingly.

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: CC @thunterdb for an opinion on that one, as he has touched most of this code.

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: @VinceShieh if you're interested in proceeding with the change you describe, go ahead. The new behavior should be documented explicitly, because I think it's the behavior one would already expect.

**Comment by srowen** [comment]
OK, I think it's also fair to just merge this as is. It's possible that later approxQuantile should be changed to ignore NaN.

**Comment by srowen** [comment]
Merged to master
